File: packages/lynx-background/ios/BackgroundPublisher.swift
import Foundation
import Lynx

/// Per-`LynxView` publisher that pumps `BackgroundEventBus` payloads into JS
/// via `LynxView.sendGlobalEvent(name, withParams:)`.
///
/// One instance per LynxView; instantiated by the generated
/// `GeneratedLifecyclePublishers.attachAll(to:)` and retained for the
/// LynxView's lifetime. The bus is global so a BGTask that fires before the
/// LynxView's JS heap is ready will queue under its `runId`; the JS shim
/// completes such a fire as `success: false` after the bounded grace period.
final class BackgroundPublisher {

    private weak var lynxView: LynxView?
    private var token: UUID?

    init(lynxView: LynxView) {
        self.lynxView = lynxView
        self.token = BackgroundEventBus.shared.addListener { [weak self] channel, payload in
            guard let view = self?.lynxView else { return }
            view.sendGlobalEvent(channel, withParams: [payload])
        }
    }

    deinit {
        if let token = token {
            BackgroundEventBus.shared.removeListener(token)
        }
    }
}

File: packages/lynx-background/android/com/sigx/background/BackgroundPublisher.kt
package com.sigx.background

import android.util.Log
import com.lynx.react.bridge.JavaOnlyArray
import com.lynx.tasm.LynxView
import java.util.UUID

/**
 * Per-[LynxView] publisher that pumps [BackgroundEventBus] payloads into JS
 * via `LynxView.sendGlobalEvent(channel, [payload])`.
 *
 * One instance per LynxView; instantiated by the generated
 * `GeneratedLifecyclePublishers.attachAll(lynxView)` and retained for the
 * LynxView's lifetime.
 */
class BackgroundPublisher(private val lynxView: LynxView) {

    private var token: UUID? = null

    fun attach() {
        token = BackgroundEventBus.addListener { channel, payload ->
            try {
                val params = JavaOnlyArray()
                params.pushMap(payload)
                lynxView.sendGlobalEvent(channel, params)
            } catch (e: Throwable) {
                Log.w(TAG, "publish failed: ${e.message}")
            }
        }
    }

    fun detach() {
        token?.let { BackgroundEventBus.removeListener(it) }
        token = null
    }

    private companion object {
        const val TAG = "SigxBackgroundPublisher"
    }
}
